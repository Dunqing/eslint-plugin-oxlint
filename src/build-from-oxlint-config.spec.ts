import { describe, expect, it } from 'vitest';
import {
  buildFromOxlintConfig,
  buildFromOxlintConfigFile,
} from './build-from-oxlint-config.js';
import fs from 'node:fs';

describe('buildFromOxlintConfig', () => {
  describe('rules values', () => {
    it('detect active rules inside "rules" scope', () => {
      ['error', ['error'], 'warn', ['warn'], 1, [1], 2, [2]].forEach(
        (ruleSetting) => {
          const rules = buildFromOxlintConfig({
            rules: {
              eqeqeq: ruleSetting,
            },
          });

          expect('eqeqeq' in rules).toBe(true);
          expect(rules.eqeqeq).toBe('off');
        }
      );
    });

    it('skip deactive rules inside "rules" scope', () => {
      ['off', ['off'], 0, [0]].forEach((ruleSetting) => {
        const rules = buildFromOxlintConfig({
          rules: {
            eqeqeq: ruleSetting,
          },
        });

        expect('eqeqeq' in rules).toBe(false);
      });
    });

    it('skip invalid rules inside "rules" scope', () => {
      ['on', ['on'], 3, [3]].forEach((ruleSetting) => {
        const rules = buildFromOxlintConfig({
          rules: {
            eqeqeq: ruleSetting,
          },
        });

        expect('eqeqeq' in rules).toBe(false);
      });
    });
  });

  it('skip deactive categories', () => {
    expect(
      buildFromOxlintConfig({
        categories: {
          // correctness is the only category on by default
          correctness: 'off',
        },
      })
    ).toStrictEqual({});
  });

  it('default plugins (react, unicorn, typescript), default categories', () => {
    // snapshot because it can change with the next release
    expect(buildFromOxlintConfig({})).toMatchSnapshot(
      'defaultPluginDefaultCategories'
    );
  });

  it('custom plugins, default categories', () => {
    // snapshot because it can change with the next release
    expect(
      buildFromOxlintConfig({
        plugins: ['unicorn'],
      })
    ).toMatchSnapshot('customPluginDefaultCategories');
  });

  it('custom plugins, custom categories', () => {
    // snapshot because it can change with the next release
    expect(
      buildFromOxlintConfig({
        plugins: ['import'],
        categories: {
          nursery: 'warn',
          correctness: 'off',
        },
      })
    ).toMatchSnapshot('customPluginCustomCategories');
  });

  it('skip deactive rules, for custom enable category', () => {
    const rules = buildFromOxlintConfig({
      plugins: ['import'],
      categories: {
        nursery: 'warn',
        correctness: 'off',
      },
      rules: {
        'import/no-unused-modules': 'off',
      },
    });
    expect('import/no-unused-modules' in rules).toBe(false);
  });
});

const createConfigFileAndBuildFromIt = (
  filename: string,
  content: string
): Record<string, unknown> => {
  fs.writeFileSync(filename, content);

  const rules = buildFromOxlintConfigFile(filename);

  fs.unlinkSync(filename);

  return rules;
};

describe('buildFromOxlintConfigFile', () => {
  it('successfully parse oxlint config', () => {
    const rules = createConfigFileAndBuildFromIt(
      'success-config.json',
      JSON.stringify({
        rules: {
          'no-await-loop': 'error',
        },
      })
    );

    expect('no-await-loop' in rules).toBe(true);
  });

  it('fails to find oxlint config', () => {
    const rules = buildFromOxlintConfigFile('not-found.json');

    expect(rules).toStrictEqual({});
  });

  it('fails to parse invalid json', () => {
    const rules = createConfigFileAndBuildFromIt(
      'invalid-json.json',
      '["this", is an invalid json format]'
    );

    expect(rules).toStrictEqual({});
  });

  it('fails to parse invalid oxlint config', () => {
    const rules = createConfigFileAndBuildFromIt(
      'invalid-config.json',
      JSON.stringify(['this is valid json but not an object'])
    );

    expect(rules).toStrictEqual({});
  });
});
