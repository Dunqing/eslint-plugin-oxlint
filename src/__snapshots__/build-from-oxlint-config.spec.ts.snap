// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`custom plugins, custom categories > customPluginCustomCategories 1`] = `
{
  "no-await-in-loop": "off",
}
`;

exports[`custom plugins, default categories > customPluginDefaultCategories 1`] = `
{
  "no-async-promise-executor": "off",
  "no-caller": "off",
  "no-class-assign": "off",
  "no-cond-assign": "off",
  "no-const-assign": "off",
  "no-constant-binary-expression": "off",
  "no-constant-condition": "off",
  "no-control-regex": "off",
  "no-delete-var": "off",
  "no-dupe-class-members": "off",
  "no-dupe-else-if": "off",
  "no-dupe-keys": "off",
  "no-duplicate-case": "off",
  "no-empty-character-class": "off",
  "no-empty-pattern": "off",
  "no-ex-assign": "off",
  "no-func-assign": "off",
  "no-global-assign": "off",
  "no-import-assign": "off",
  "no-invalid-regexp": "off",
  "no-irregular-whitespace": "off",
  "no-loss-of-precision": "off",
  "no-new-native-nonconstructor": "off",
  "no-obj-calls": "off",
  "no-self-assign": "off",
  "no-setter-return": "off",
  "no-shadow-restricted-names": "off",
  "no-sparse-arrays": "off",
  "no-this-before-super": "off",
  "no-unsafe-finally": "off",
  "no-unsafe-optional-chaining": "off",
  "no-unused-private-class-members": "off",
  "no-useless-catch": "off",
  "no-useless-rename": "off",
  "no-with": "off",
  "require-yield": "off",
  "unicorn/no-await-in-promise-methods": "off",
  "unicorn/no-document-cookie": "off",
  "unicorn/no-empty-file": "off",
  "unicorn/no-invalid-remove-event-listener": "off",
  "unicorn/no-thenable": "off",
}
`;

exports[`default plugins (react, unicorn, typescript), default categories > defaultPluginDefaultCategories 1`] = `
{
  "@typescript-eslint/no-dupe-class-members": "off",
  "@typescript-eslint/no-duplicate-enum-values": "off",
  "@typescript-eslint/no-extra-non-null-assertion": "off",
  "@typescript-eslint/no-loss-of-precision": "off",
  "@typescript-eslint/no-misused-new": "off",
  "@typescript-eslint/no-non-null-asserted-optional-chain": "off",
  "@typescript-eslint/no-this-alias": "off",
  "@typescript-eslint/no-unsafe-declaration-merging": "off",
  "@typescript-eslint/triple-slash-reference": "off",
  "no-async-promise-executor": "off",
  "no-caller": "off",
  "no-class-assign": "off",
  "no-cond-assign": "off",
  "no-const-assign": "off",
  "no-constant-binary-expression": "off",
  "no-constant-condition": "off",
  "no-control-regex": "off",
  "no-delete-var": "off",
  "no-dupe-class-members": "off",
  "no-dupe-else-if": "off",
  "no-dupe-keys": "off",
  "no-duplicate-case": "off",
  "no-empty-character-class": "off",
  "no-empty-pattern": "off",
  "no-ex-assign": "off",
  "no-func-assign": "off",
  "no-global-assign": "off",
  "no-import-assign": "off",
  "no-invalid-regexp": "off",
  "no-irregular-whitespace": "off",
  "no-loss-of-precision": "off",
  "no-new-native-nonconstructor": "off",
  "no-obj-calls": "off",
  "no-self-assign": "off",
  "no-setter-return": "off",
  "no-shadow-restricted-names": "off",
  "no-sparse-arrays": "off",
  "no-this-before-super": "off",
  "no-unsafe-finally": "off",
  "no-unsafe-optional-chaining": "off",
  "no-unused-private-class-members": "off",
  "no-useless-catch": "off",
  "no-useless-rename": "off",
  "no-with": "off",
  "react/jsx-key": "off",
  "react/jsx-no-duplicate-props": "off",
  "react/jsx-no-target-blank": "off",
  "react/jsx-no-undef": "off",
  "react/no-children-prop": "off",
  "react/no-danger-with-children": "off",
  "react/no-direct-mutation-state": "off",
  "react/no-find-dom-node": "off",
  "react/no-is-mounted": "off",
  "react/no-render-return-value": "off",
  "react/no-string-refs": "off",
  "react/void-dom-elements-no-children": "off",
  "require-yield": "off",
  "unicorn/no-await-in-promise-methods": "off",
  "unicorn/no-document-cookie": "off",
  "unicorn/no-empty-file": "off",
  "unicorn/no-invalid-remove-event-listener": "off",
  "unicorn/no-thenable": "off",
}
`;
